cyralinc/terraform-provider-cyral
# ENG-9174: Remove deprecated resources for major version 3

## Description of the change

The following resources were removed:

- `cyral_repository_identity_map`(Removed in #293)
- `cyral_repository_local_account`(Removed in #293)
- `cyral_datamap`
- `cyral_identity_map` (Removed in #293)
- `cyral_integration_sso_aad`
- `cyral_integration_sso_adfs`
- `cyral_integration_sso_forgerock`
- `cyral_integration_sso_gsuite`
- `cyral_integration_sso_okta`
- `cyral_integration_sso_ping_one`

Removed arguments in resource schemas:

- `cyral_policy.properties`

Removed arguments in provider schema:

- `auth_provider`
- `auth0_audience`
- `auth0_domain`
- `auth0_client_id`
- `auth0_client_secret`

(Use of keycloak should be transparent.)

`scripts` folder was removed and references to it in the docs were removed.

State migration for was added for following resources, to fix the format of the `id` argument:

- `cyral_repository_binding` -- was `{sidecar_id}-{repository_id}`, now `{sidecar_id}/{repository_id}`
- `cyral_repository_conf_auth` -- was `repo-conf`, now `{repository_id}`
- `cyral_repository_conf_analysis` -- was `{repository_id}/ConfAnalysis`, now `{repository_id}`
- `cyral_policy_rule` -- was `{policy_rule_id}`, now `{policy_id}/{policy_rule_id}`
- `cyral_role_sso_groups` -- was `{role_id}/SSOGroups`, now `{role_id}`

## Type of change


## Checklists

### Development


### Code review

- [x] Jira issue referenced in commit message and/or PR title

### Testing

#### Unit tests
Unit tests were added to test state migration of the resources:
- `cyral_role_sso_groups`
- `cyral_repository_conf_analysis`
- `cyral_repository_conf_auth`
- `cyral_repository_binding`
- `cyral_policy_rule`

The state migration is limited to the `id` argument of each resource. Previously, these resources had weird IDs. For `cyral_policy_rule`, I added a new argument `policy_rule_id` so that the user can directly access the policy rule ID in case it is necessary.

#### Acceptance tests
Acceptance tests are passing.

#### Manual Testing
For each resource below, I performed a `terraform apply` using provider version `v2.9.0`. Then, I changed the provider to the local provider pointing to this branch (i.e. `local/terraform/cyral`). Then, I hit `terraform show` to show the resource configuration before the update. Then, I hit `terraform refresh` or `terraform apply`, and then `terraform show` again to show the resource configuration. The before and after of relevant state migration resource configurations are shown below. For each of them, I made sure to run `terraform apply` after the upgrade to make sure it did not start failing or produce weird plans.

**Role SSO Groups -- Before**
```terraform
# cyral_role_sso_groups.t1:
resource "cyral_role_sso_groups" "t1" {
    id      = "361cf35b-9978-4022-9302-b85fe28f1169/SSOGroups"
    role_id = "361cf35b-9978-4022-9302-b85fe28f1169"

    sso_group {
        group_name = "Everyone"
        id         = "81537368-1066-4b6c-8a31-7dcb4872cac4"
        idp_id     = "okta.1d370154-a721-4d17-9b5e-e8c69778640c"
        idp_name   = "Okta"
    }
}
```

**Role SSO Groups -- After**
```terraform
# cyral_role_sso_groups.t1:
resource "cyral_role_sso_groups" "t1" {
    id      = "361cf35b-9978-4022-9302-b85fe28f1169"
    role_id = "361cf35b-9978-4022-9302-b85fe28f1169"

    sso_group {
        group_name = "Everyone"
        id         = "81537368-1066-4b6c-8a31-7dcb4872cac4"
        idp_id     = "okta.1d370154-a721-4d17-9b5e-e8c69778640c"
        idp_name   = "Okta"
    }
}
```


**Repository conf analysis -- Before**
```terraform
# cyral_repository_conf_analysis.t1:
resource "cyral_repository_conf_analysis" "t1" {
    alert_on_violation            = true
    block_on_violation            = false
    disable_filter_analysis       = false
    disable_pre_configured_alerts = false
    enable_data_masking           = false
    id                            = "2FM1ZUZOgMJy6q2S5DeOV1XGEkH/ConfAnalysis"
    redact                        = "all"
    repository_id                 = "2FM1ZUZOgMJy6q2S5DeOV1XGEkH"
    rewrite_on_violation          = false
}
```

**Repository conf analysis -- After**
```terraform
# cyral_repository_conf_analysis.t1:
resource "cyral_repository_conf_analysis" "t1" {
    alert_on_violation            = true
    block_on_violation            = false
    comment_annotation_groups     = []
    disable_filter_analysis       = false
    disable_pre_configured_alerts = false
    enable_data_masking           = false
    id                            = "2FM1ZUZOgMJy6q2S5DeOV1XGEkH"
    log_groups                    = []
    redact                        = "all"
    repository_id                 = "2FM1ZUZOgMJy6q2S5DeOV1XGEkH"
    rewrite_on_violation          = false
}
```

**Repository conf auth -- Before**

```terraform
# cyral_repository_conf_auth.t1:
resource "cyral_repository_conf_auth" "t1" {
    allow_native_auth = false
    client_tls        = "disable"
    id                = "repo-conf"
    repo_tls          = "disable"
    repository_id     = "2FM1JpQqWhsVVS5OaUuF2saP3jf"
}
```

**Repository conf auth -- After**
```terraform
# cyral_repository_conf_auth.t1:
resource "cyral_repository_conf_auth" "t1" {
    allow_native_auth = false
    client_tls        = "disable"
    id                = "2FM1JpQqWhsVVS5OaUuF2saP3jf"
    repo_tls          = "disable"
    repository_id     = "2FM1JpQqWhsVVS5OaUuF2saP3jf"
}
```

**Repository binding -- Before**

```terraform
# cyral_repository_binding.t1:
resource "cyral_repository_binding" "t1" {
    enabled                       = true
    id                            = "2FM0sS28tQAaMK8iA37JdoV4Mq0-2FM12IhuCWc2JaVChb8WUG5D8uP"
    listener_host                 = "0.0.0.0"
    listener_port                 = 1433
    repository_id                 = "2FM12IhuCWc2JaVChb8WUG5D8uP"
    sidecar_as_idp_access_gateway = false
    sidecar_id                    = "2FM0sS28tQAaMK8iA37JdoV4Mq0"
}
```

**Repository binding -- After**

```terraform
# cyral_repository_binding.t1:
resource "cyral_repository_binding" "t1" {
    enabled                       = true
    id                            = "2FM0sS28tQAaMK8iA37JdoV4Mq0/2FM12IhuCWc2JaVChb8WUG5D8uP"
    listener_host                 = "0.0.0.0"
    listener_port                 = 1433
    repository_id                 = "2FM12IhuCWc2JaVChb8WUG5D8uP"
    sidecar_as_idp_access_gateway = false
    sidecar_id                    = "2FM0sS28tQAaMK8iA37JdoV4Mq0"
```

**Policy rule -- Before**
```terraform
# cyral_policy_rule.t1:
resource "cyral_policy_rule" "t1" {
    id        = "2FM0P24c0BhioRNrgSP6tQqlRKC"
    policy_id = "2FM0OwmIhJ31YMFcMOEh6VazrH5"
}
```

**Policy Rule -- After**
```terraform
# cyral_policy_rule.t1:
resource "cyral_policy_rule" "t1" {
    hosts          = []
    id             = "2FM0OwmIhJ31YMFcMOEh6VazrH5/2FM0P24c0BhioRNrgSP6tQqlRKC"
    policy_id      = "2FM0OwmIhJ31YMFcMOEh6VazrH5"
    policy_rule_id = "2FM0P24c0BhioRNrgSP6tQqlRKC"
}
```
